src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -321,6 +328,206 @@ public async Task<GetTransactionModel> GetTransactionAsync(string txid)
             return model;
         }
 
+        [ActionName("listaddressgroupings")]
+        [ActionDescription("Returns a list of grouped addresses which have had their common ownership made public by common use as inputs or as the resulting change in past transactions.")]
+        public async Task<AddressGroupingModel[]> ListAddressGroupings(string walletName, CoinType coinType)
+        {
+            //Get the wallet to check.
+            var wallet = this.walletManager.GetWallet(walletName);
+
+            //Get all the address groupings.
+            var addressGroupings = await GetAddressGroupingsAsync(wallet, coinType);
+
+            var addressGroupingModels = new List<AddressGroupingModel>();
+
+            foreach (var addressGrouping in addressGroupings)
+            {
+                foreach (var address in addressGrouping)
+                {
+                    var balance = this.walletManager.GetAddressBalance(address.ToString());
+                    var addressGroupingModel = new AddressGroupingModel()
+                    {
+                        Address = BitcoinAddress.Create(address.ToString(), this.Network),
+                        Amount = balance.AmountConfirmed
+                    };
+
+                    addressGroupingModels.Add(addressGroupingModel);
+                }
+            }
+
+            return addressGroupingModels.ToArray();
+        }
+
+        /// <summary>
+        /// Please see https://github.com/bitcoin/bitcoin/blob/726d0668ff780acb59ab0200359488ce700f6ae6/src/wallet/wallet.cpp#L3641
+        /// </summary>
+        /// <param name="wallet">The wallet to check.</param>
+        /// <param name="coinType">The cointype for this wallet.</param>
+        /// <returns>A list of base 58 addresses.</returns>
+        private async Task<List<List<BitcoinPubKeyAddress>>> GetAddressGroupingsAsync(Wallet wallet, CoinType coinType)
+        {
+            //Get all the transaction data records for this wallet.
+            var transactions = wallet.GetAllTransactionsByCoinType(coinType);
+
+            var groupings = new List<List<BitcoinPubKeyAddress>>();
+
+            foreach (var transaction in transactions)
+            {
+                var tx = await this.blockStore.GetTransactionByIdAsync(transaction.Id);
```

Comment:
If `-txindex=1` isn't specified should a more user-friendly error be returned than potentially returning nothing?

Reply:
I will now check transaction indexing when the method is called.